### PR TITLE
[ExportVerilog] Remove getValue for InstanceOp

### PR DESCRIPTION
**Pull request by prithayan**
Remove the usage of `getValue` since it recomputes the port list for every port. This is very expensive when there are huge number of ports.

### CODE REVIEW

**Comment by trilorez** [comment]
Looks like there's another instance of `getValue` [here](https://github.com/llvm/circt/blob/77bd8e388066a03eff801f0dbf02dc40fd7e304e/lib/Conversion/ExportVerilog/ExportVerilog.cpp#L5188).

I think if you handle that one too, you can remove `InstanceOp::getValue` entirely.

**Review by darthscsi — COMMENTED** [review]
Try this:
```
diff --git a/lib/Dialect/HW/HWOps.cpp b/lib/Dialect/HW/HWOps.cpp
index 80e9c7e0d..ffab2093a 100644
--- a/lib/Dialect/HW/HWOps.cpp
+++ b/lib/Dialect/HW/HWOps.cpp
@@ -1732,16 +1732,9 @@ size_t InstanceOp::getPortIdForOutputId(size_t idx) {
 }
 
 Value InstanceOp::getValue(size_t idx) {
-  auto mpi = getPortList();
-  size_t inputPort = 0, outputPort = 0;
-  for (size_t x = 0; x < idx; ++x)
-    if (mpi.at(x).isOutput())
-      ++outputPort;
-    else
-      ++inputPort;
-  if (mpi.at(idx).isOutput())
-    return getResults()[outputPort];
-  return getInputs()[inputPort];
+  if (idx < getNumInputPorts())
+    return getInputs()[idx];
+  return getResults()[idx - getNumInputPorts()];
 }
 
 //===----------------------------------------------------------------------===//
 ```

File: lib/Dialect/HW/HWOps.cpp [hunk 1]
```diff
@@ -1733,6 +1734,10 @@ size_t InstanceOp::getPortIdForOutputId(size_t idx) {
 
 Value InstanceOp::getValue(size_t idx) {
   auto mpi = getPortList();
+  return getValue(idx, mpi);
+}
+
+Value InstanceOp::getValue(size_t idx, ModulePortInfo &mpi) {
```
Comment (uenoku) [inline lib/Dialect/HW/HWOps.cpp:1740]: Does this fix the original regression? It looks it takes O(N^2) still (maybe I'm wrong).

**Comment by prithayan** [comment]
Few things that we need to consider here, 
1. The module ports may not be ordered input-output, so we need to iterate over the `ModulePortInfo` to compute the input/output port index from the module port index.
2. Computing the `ModulePortInfo` inside the `getValue` is very expensive.
3. The port mapping logic shouldn't leak into the client. So we shouldn't just inline the `getValue`.
The final proposed solution here is to pre-compute the module port index to the instance Values map. The only cost here is the additional storage and its initialization to store the Values.

**Review by trilorez — APPROVED** [review]
Another approach is to make a custom iterator that computes the port values, which would avoid needing to allocate storage for all the values. But it's probably not necessary.

File: include/circt/Dialect/HW/HWStructure.td [hunk 2]
```diff
@@ -493,8 +493,10 @@ def InstanceOp : HWOp<"instance", [
     Operation *getReferencedModule(SymbolTable& tbl);
     Operation *getReferencedModuleSlow();
 
-    /// Return the value for a port in port order.
-    Value getValue(size_t idx);
+    /// Return the values for the port in port order.
+    /// Note: The module ports maynot be input,output ordered. This computes
+    /// the port index to instance result/input Value mapping.
+    void  getValues(SmallVectorImpl<Value> &values, const ModulePortInfo &mpi);
```
Comment (trilorez) [inline include/circt/Dialect/HW/HWStructure.td:499]: Nit: extra space between `void` and `getValues`

File: include/circt/Dialect/HW/HWStructure.td [hunk 3]
```diff
@@ -493,8 +493,10 @@ def InstanceOp : HWOp<"instance", [
     Operation *getReferencedModule(SymbolTable& tbl);
     Operation *getReferencedModuleSlow();
 
-    /// Return the value for a port in port order.
-    Value getValue(size_t idx);
+    /// Return the values for the port in port order.
+    /// Note: The module ports maynot be input,output ordered. This computes
```
Comment (trilorez) [inline include/circt/Dialect/HW/HWStructure.td:497]: Nit: no space in `maynot` and `input,output`

**Comment by prithayan** [comment]
> Another approach is to make a custom iterator that computes the port values, which would avoid needing to allocate storage for all the values. But it's probably not necessary.

Good idea, I will merge this now, so that it can be included in the next firtool bump tomorrow. We can explore the iterator implementation to avoid the storage.

**Comment by prithayan** [comment]
Thanks for the review everyone.

**Pull request merged**